File: desktop/src/main/java/org/gipsybuho/recetasfamiliares/data/repository/ShoppingListRepository.java
package org.gipsybuho.recetasfamiliares.data.repository;

import org.gipsybuho.recetasfamiliares.api.ApiClient;
import org.gipsybuho.recetasfamiliares.api.ApiException;
import org.gipsybuho.recetasfamiliares.api.dto.SyncDtos;
import org.gipsybuho.recetasfamiliares.core.AppSession;

import java.util.List;

public class ShoppingListRepository {

    private final ApiClient api;
    private final AppSession session;

    public ShoppingListRepository(ApiClient api, AppSession session) {
        this.api = api;
        this.session = session;
    }

    public List<SyncDtos.ShoppingDtos.ShoppingListDto> loadAll() throws ApiException {
        String familyId = session.getFamilyId();
        String path = "api/v1/families/" + familyId + "/shopping-lists?page=0&size=50";
        SyncDtos.ShoppingDtos.ShoppingPageResponse page =
                api.get(path, SyncDtos.ShoppingDtos.ShoppingPageResponse.class);
        if (page.items() == null) return List.of();
        return page.items().stream().filter(l -> !l.deleted()).toList();
    }

    public List<SyncDtos.ShoppingDtos.ShoppingListItemDto> loadItems(String listId) throws ApiException {
        String familyId = session.getFamilyId();
        // El backend limita size a 100 (@Max); 200 provocaba un 400 silencioso.
        String path = "api/v1/families/" + familyId + "/shopping-lists/" + listId + "/items?page=0&size=100";
        SyncDtos.ShoppingDtos.ShoppingItemPageResponse page =
                api.get(path, SyncDtos.ShoppingDtos.ShoppingItemPageResponse.class);
        if (page.items() == null) return List.of();
        return page.items().stream().filter(i -> !i.deleted()).toList();
    }

    public SyncDtos.ShoppingDtos.ShoppingListItemDto checkItem(
            SyncDtos.ShoppingDtos.ShoppingListItemDto item, boolean checked) throws ApiException {
        String familyId = session.getFamilyId();
        var req = new SyncDtos.ShoppingDtos.UpdateShoppingListItemRequest(
                item.position(), item.name(), item.quantity(), item.unit(), checked, item.note());
        String path = "api/v1/families/" + familyId
                + "/shopping-lists/" + item.shoppingListId()
                + "/items/" + item.id();
        return api.put(path, req, SyncDtos.ShoppingDtos.ShoppingListItemDto.class);
    }

    public void updateFromSync(
            List<SyncDtos.ShoppingDtos.ShoppingListDto> lists,
            List<SyncDtos.ShoppingDtos.ShoppingListItemDto> items
    ) {
        // ShoppingListRepository no mantiene cache local: listas e items se cargan bajo demanda.
    }
}

File: ios/composeApp/src/commonMain/kotlin/org/gipsybuho/recetasfamiliares/shopping/ShoppingListRepository.kt
package org.gipsybuho.recetasfamiliares.shopping

import io.ktor.client.call.*
import io.ktor.client.request.*
import org.gipsybuho.recetasfamiliares.core.SessionStore
import org.gipsybuho.recetasfamiliares.network.ApiClient
import org.gipsybuho.recetasfamiliares.network.PageDto
import org.gipsybuho.recetasfamiliares.network.ShoppingListDto
import org.gipsybuho.recetasfamiliares.network.ShoppingListItemDto

class ShoppingListRepository(
    private val apiClient: ApiClient,
    private val session: SessionStore
) {
    suspend fun loadLists(page: Int = 0, size: Int = 20): List<ShoppingListDto> {
        val familyId = session.familyId ?: return emptyList()
        val response: PageDto<ShoppingListDto> = apiClient.http
            .get("api/v1/families/$familyId/shopping-lists") {
                parameter("page", page)
                parameter("size", size)
            }.body()
        return response.items.filter { !it.deleted }
    }

    suspend fun loadItems(listId: String, page: Int = 0, size: Int = 100): List<ShoppingListItemDto> {
        val familyId = session.familyId ?: return emptyList()
        val response: PageDto<ShoppingListItemDto> = apiClient.http
            .get("api/v1/families/$familyId/shopping-lists/$listId/items") {
                parameter("page", page)
                parameter("size", size)
            }.body()
        return response.items.filter { !it.deleted }.sortedBy { it.position }
    }
}
